src/refs.c

```diff
@@ -1430,3 +1430,27 @@ const char *git_reference_shorthand(const git_reference *ref)
 {
 	return git_reference__shorthand(ref->name);
 }
+
+int git_reference__is_unborn_head(bool *unborn, const git_reference *ref, git_repository *repo)
+{
+	int error;
+	git_reference *tmp_ref;
+	assert(unborn && ref && repo);
+
+	if (ref->type == GIT_REF_OID) {
+		*unborn = 0;
+		return 0;
+	}
+
+	error = git_reference_lookup_resolved(&tmp_ref, repo, ref->name, -1);
+	git_reference_free(tmp_ref);
+
+	if (error != 0 && error != GIT_ENOTFOUND)
+		return error;
+	else if (error == GIT_ENOTFOUND && git__strcmp(ref->name, GIT_HEAD_FILE) == 0)
+		*unborn = true;
+	else
+		*unborn = false;
```

Comment:
Just making sure that I understand the decisions from #4780.  I'm thinking that at present that if we encounter a broken symref that's _not_ `HEAD` that we will return that unborn is false?    Should this block look like:

```
if (error == GIT_ENOTFOUND && git__strcmp(ref->name, GIT_HEAD_FILE) == 0)
    *unborn = true;
else if (!error)
    *unborn = false;

return error;
```

(Ignoring the style around ifs and such, just trying to make sure that I understand the logic.)

Reply:
Yes, that's true, a broken symref whose name is not `HEAD` will not be considered unborn, and will not error (at least here). This is the check we're doing `git_repository_head` BTW, with the `HEAD` check made more explicit.

I find your proposed block harder to reason about w.r.t to `error`, but I think it still returns errors for cases that would be fine (the 1st condition would still be `GIT_ENOTFOUND` at least). Something like this, if the inline return is not okay :

```c
if (error == 0)
    *unborn = false;
else if (error == GIT_ENOTFOUND && git__strcmp(ref->name, GIT_HEAD_FILE) == 0) {
    *unborn = true;
    error = 0;
} else
    /* error will fallthrough */
```

Your version is a step after this one, so now I get it 😉. Yours is only missing the `error = 0` in the "unborn" handler.

Reply:
Yes, you're very much correct, I was missing that error reset.